jena-arq/src/main/java/org/apache/jena/sparql/expr/NodeValue.java

```diff
@@ -783,6 +772,22 @@ private static int compare(NodeValue nv1, NodeValue nv2, boolean sortOrderingCom
                     return Expr.CMP_GREATER ;
                 return Expr.CMP_EQUAL;  // Both plain or both xsd:string.
             }
+            case VSPACE_SORTKEY :
+            {
+                int cmp = 0;
+                String c1 = nv1.getCollation();
+                String c2 = nv2.getCollation();
+                if (c1 != null && c2 != null && c1.equals(c2)) {
+                    // locales are parsed. Here we could think about caching if necessary
+                    Locale desiredLocale = Locale.forLanguageTag(c1);
+                    // collators are already stored in a concurrent map by the JVM, with <locale, softref<collator>>
+                    Collator collator = Collator.getInstance(desiredLocale);
+                    cmp = collator.compare(nv1.getString(), nv2.getString());
+                } else {
```

Comment:
Would it be better to have `NodeSortKey` being comparable rather than putting the comparison here?

The then removes the need for `getCollation`.

And, as a general mechanism, `NodeSortKey` are not restricted to language-based collection.  Maybe an extension is for `NodeSortKey` based on an enum-like interpretation of a value (e.g. a persons title or rank or even "one", "two", "three").